### PR TITLE
Mhp30

### PR DESCRIPTION
**Pull request by Ralim**
Tracking progress in bringing in support for the MHP30.
Cant merge / usable until all a ticked off.

- [x] OLED display working
- [x] Buttons
- [50%] USB-PD
- [x] Heater control
- [ ] Cold junction temp sensor
- [x] VIN measurement
- [50%] Accelerometer  (Reading but needs logic)
- [ ] Inverted logic for accelerometer knock over == stop
- [ ] How do we handle bootloader (do we need to do our own)
- [ ] RGB LED
- [ ] Buzzer

### CODE REVIEW

**Comment by Ralim** [comment]
Alrighty this is good to merge I believe.
If I dont hear any reason not to, ill merge this in, in about a day or so :)

**Comment by Vinigas** [comment]
Flash successful.
Notes:
o Orientation setting "Automatic" is not working
o Not heating then plugged in into PC 5V 3.5W (maybe heat dissipated faster than it is heating)
o When plugged into PD 45W firmware, it visibly crashes in screen and RGB (like safely short-circuited) and restarts itself back to main menu.
o Power limit setting: cannot set 65W, because it is stepping in 2W, only 64W can be set.

**Comment by Ralim** [comment]
> Flash successful.
> Notes:
> o Orientation setting "Automatic" is not working

Ah right, I need to suppress orientation via accelerometer.

> o When plugged into PD 45W (TS80P official charger), and pressing A to heating, it visibly crashes 1s in screen and RGB (like safely glitched) and restarts itself back to main menu.

Your charger is not rated for 60W+. 
Support below 60W will depend on how fussy your charger is.
I'll have another quick look to see if i can improve this.

> o Power limit setting: cannot be set 65W by user, because it is stepping in 2W, only 64W can be set.

Ah heck, thats a fun bug.

> o Car 12V QC3 only works 5V3W

The MHP30 does not support QC. So you will only negotiate 5V.

**Comment by Ralim** [comment]
> o When plugged into PD 45W (TS80P official charger), and pressing A to heating, it visibly crashes 1s in screen and RGB (like safely glitched) and restarts itself back to main menu.

@Vinigas 
Could you try setting the power limit to 40W and re-testing?
Looking here the inductor appears to be doing it thing and the wattage limiter looks to work.
So I may just need to back it off by a watt or two

**Comment by Ralim** [comment]
@Vinigas 

I think this should be good to re-test.
Added a 1W limit backoff from the power supply max.
On all of my power banks here (only non 65W+ things I have 😢 )
Looking on my power supply plot it looks fairly clean current wise.. :/

So would be curious if you could give it a test again.
If it still resets, you could try dropping the power limiter down.

@g3gg0
Could you also give it a go :)

**Comment by Vinigas** [comment]
When plugged into PD 45W (TS80P official charger), I need to set power-limit 13W. I used binary search to find this value.

But what is strange that on first try 18W power limit worked. And after that only 13W limit, after that 14W limit worked too. After more tries. Looks like it is very related that hotplate is cold then heating begins and temperature is set low then heating starts. Firmware crashes then wattage raises. The faster it raises, the faster crash.

Then testing I always set temperature below 100'C. Usually 50'C or 80'C. Maybe wattage is increase too sharply, then target temperature is near current temperature (especially on heating start) while hotplace it cold.

 Just shallowly theorizing...

**Comment by Ralim** [comment]
Do you have any details on the TS80P official charger? (I don't have one as no AU plug last I checked).

Not surprised if we are running into this, the official firmware inserts a dead time every so often in their heater drive where we this firmware doesn't. Which probably helps lower the average higher current. Though my quick testing made the current look fairly uniform 😓

**Comment by Vinigas** [comment]
image
image
image

**Comment by Ralim** [comment]
Spam is never an issue really :)
When those builds finish, can you give them a go? Added a slew rate limit for ramping up on power + fixed a bug with the limiter I think as well
Not quite pushing to the limit, still like 1.5-2W below limit by my approx measure but hoping it helps

**Comment by Ralim** [comment]
I found a charger that tripped off on old firmware that doesnt on new, so hoping you have improved results.

**Comment by Vinigas** [comment]
Will keep updating this comment with various power-limits values on this chargers:
o 18W haven't died.
o 65W died
o 30W died (crashed then 23.4W usage was showing)
o ...

**Comment by Ralim** [comment]
That is really weird.
Just to be sure could you try a settings reset to rule it out :)

I can't reproduce this here at the moment, and it makes no sense really assuming the charger can actually do 45W continuously.

I'll try slowing it down more tomorrow

**Comment by Vinigas** [comment]
Seems like factory reset worked (!?!?)

**Comment by Vinigas** [comment]
image
With this 65W charger everything works, no bugs. Limiter 65W.

**Comment by Ralim** [comment]
Ahh, I know what it is.
Its happening when I drop the timer speed down to play the tone on the speaker.
As its probably dropping down too far and the current spike is getting through.
Ill re-test tonight for you

**Comment by Ralim** [comment]
@Vinigas 

I gave up trying to make the timing work out, so just masking heating for the 0.3 seconds for now.
Its either that or banging the tone, which feels about as hackish :/

Feel like a re-test :)

**Comment by Vinigas** [comment]
Re-tested.

Doesn't crash on any power supplies. And that's what the most important on first support release, in my opinion.

Seems very great. Works stable. 😊

---

Some notes for _future_ releases:
o RGB coloring is not consistent throughout the menu (heating and main-menu changes color). In default software color always showed approximate temperature of hotplate. Maybe settings could allow to set temperature in #FFFFFF format for user set temperature ranges (dynamic ranges, dynamic amount of ranges).

**Comment by Vinigas** [comment]
Found some bug just now with long pressing BOOST. Cannot tell exactly what it was, but then hotplate was on target temp and pressed boost, temp rised and then beeper screemed non stop, till temp dropped slowly to targettemp+10'C.

Wasn't able to replicate it once more. I may test more. Anyways, it doesn't crash and works fine.

**Comment by g3gg0** [comment]
temperatures / color:

i personally prefer smooth transitions to signal the color. have implemented this on my 3D printer
faint blue: cold (< 30°C) and no heating
(>= 30°C or heating) switch to orange and fade over yellow to white till 260°C
over 260°C red warning signal because exceeding specification

```
struct sColorRgb
{
  uint8_t r;
  uint8_t g;
  uint8_t b;
};

struct sColorEntry
{
  short temp;
  struct sColorRgb color;
};

struct sColorEntry colorTable[] = {
  {   0, { 10, 10, 10} },
  {  10, { 10, 10, 10} },
  {  11, {  0,  0, 40} },
  {  30, {  0,  0, 40} },
  {  60, { 60, 20,  0} },
  { 180, {100, 50,  0} },
  { 260, {100,100,100} },
  { 290, {100,100,100} },
  { 295, {100,  0,  0} },
  { 999, {100,  0,  0} }  
};

uint32_t calcColor(float celsius)
{
  uint32_t color = 0;

  for(int pos = 1; pos < COUNT(colorTable); pos++)
  {
    if(colorTable[pos].temp > celsius)
    {
      struct sColorEntry *cur = &colorTable[pos-1];
      struct sColorEntry *next = &colorTable[pos];

      struct sColorRgb baseRgb = cur->color;
      struct sColorRgb nextRgb = next->color;

      float distance = (celsius - cur->temp) / (next->temp - cur->temp);

      /* this is in PWM 0-255 */
      uint8_t r = 0;
      uint8_t g = 0;
      uint8_t b = 0;

      channelFade(&r, baseRgb.r, nextRgb.r, distance);
      channelFade(&g, baseRgb.g, nextRgb.g, distance);
      channelFade(&b, baseRgb.b, nextRgb.b, distance);
      
      return strip.Color(r, g, b);
    }
  }
  
  return strip.Color(255, 0, 0);
}
```

**Comment by Ralim** [comment]
Led colour is something I'm planning to leave out of this current PR, and welcome the community to provide PR's on how its done.

Personally I'm fairly indifferent to the LED other than wanting bright colours = burn risk. (Prefer not white but not hard rule on that) 

The crashes going into and out of boost mode I've only been able to replicate once and that was on my 30W adaptor so I'm more suspect of power glitching at this point :/

I think I'll merge the current status of the PR so I can  include this in the next release, and then ask that we split out separate issues for each improvement to keep the noise down :)

So if you could split out issues for remaining issues would be fantastic, as I think we are now beyond usable and into refinement stage.

**Comment by Ralim** [comment]
@g3gg0

In terms of your colour code, should be easy enough to add something like that when I have time (or PR welcome).

Current implementation is mostly MVP :)

**Comment by Vinigas** [comment]
image
I think colors should transition similar to this, and the hottest color being white.

Smooth color transitioning would be very cool.

**Comment by g3gg0** [comment]
ah also posted in the other thread something about the temperatures.
prefer blue, yellow, orange, white
and red for "error states" while green could some kind of "ready" signal

**Comment by g3gg0** [comment]
@Ralim 
just want to say a big fat 
### thanks for your work!

**Comment by Ralim** [comment]
Hehe, I think we will end up needing settings on this already 😁 since to me white = on, green = safe, blue =cold and red = hot

🤣

**Comment by g3gg0** [comment]
in the end the user gets used to the colors anyway, no matter which got picked ;)

**Comment by Vinigas** [comment]
As this project grows, I think that download artifacts should be not firmware, but firmware installer/compiler, which build user preferred modules (features) and compiler would look at user editable config file (like linux kernel smth).

User downloads portable IronOS builder, launches it, checks what he wants and how, then custom firmware is built. This builder would allow to select language (if GUI) etc. This innovation path would allow to save hardware ROM for features user wants as project grows in features more than hardware can contain.

E.g. if user want full Tetris, he can even uncheck compiler "add feature" for heating the tip. 😂

**Comment by g3gg0** [comment]
(added a graphic above, visualizing the coloring i meant)

regarding building:
i suspect that the average john doe user just wants to download the firmware.
even me. i prefer downloading the binaries, not wanting to even flip a bit in the hex image.
definitely not building from source. this iron/hotplate is a tool i wanted to use :D

_however_ you could provide a separate hex with the settings area that could be customized by the user.
not sure if the DFU mode could handle that properly, but having a small C/.py tool that
generates a maybe 128 byte .hex from out of nowhere that flashes into the settings area.
that is soem concept which is very common, especially in automotive area.

**Comment by Ralim** [comment]
My general feeling on the compiler tool is that most people don't really want to deal with that, plus packaging compilers for multiple distributions is a nightmare.

I haven't come across a single electron app that is smaller than the zip of firmwares to date 😂 also my general preference is anything but electron. (When electron stops shipping me basically a browser to run a Web page maybe I'll be interested). 

I agree a config tool / page that generates a hex to flash in the settings area is a better way to go, but also that feels largely a nightmare for support.
One could technically write it _now_ as all you need to do is repeat the content a bunch of times as DFU has a ~4k min file size.

I've thought about making up a qt or fyne gui to do this but just don't have the time at all.

My plan was the tool does:

1. You select your device
2. Select release
3. It downloads the zip and extracts this
4. It configures bootlogo, settings and firmware into merged hex
5. It wraps up flashing this as well


But building that out is a chunk of time 😢

**Comment by g3gg0** [comment]
> I agree a config tool / page that generates a hex to flash in the settings area is a better way to go, but also that feels largely a nightmare for support.

ok then we would have to reserve 4k of the flash for the settings area?
What about a simple web tool, maybe javascript, where you just download the configuration hex after configuring the settings?
building a hex isnt magic and the basics would be done within a day or two.

**Comment by Ralim** [comment]
Ah, no we only need to reserve one erase page (1k), you literally copy-paste the hex file lines to make the resulting hex file larger than 4k.

No firmware changes required really.

Yeah Web page is definitely a valid way to go about this, only magic is knowing settings offset based on model number.

**Comment by g3gg0** [comment]
> settings offset based on model number

cant we use a fixed location in the flash for the future? or do we have different chips?

**Comment by Ralim** [comment]
Different sized flash and we place settings in one of the last pages to avoid dealing with linking over a gap

**Comment by g3gg0** [comment]
just googled a BSD 3-clause hex handling library that could fit:
https://github.com/NordicSemiconductor/nrf-intel-hex

**Comment by g3gg0** [comment]
linking over a gap shouldnt be an issue, or?

**Comment by Ralim** [comment]
Shouldn't be an issue, but ran into issues in my quick attempt to get it to work nicely.

Also given we have bumped into 0% free flash a few times already in the history of the firmware, concern around packing density against aligning to the gap is a thing.

But yeah I agree it could definitely be done, most likely by just careful linker set up.

**Comment by g3gg0** [comment]
yeah understand and agree. reserving 1k ist tough anyway.

**Comment by Zuckme** [comment]
> temperatures / color:
> 
> i personally prefer smooth transitions to signal the color. have implemented this on my 3D printer faint blue: cold (< 30°C) and no heating (>= 30°C or heating) switch to orange and fade over yellow to white till 260°C over 260°C red warning signal because exceeding specification
> 
> thats something even my kids understood :)
> 
> ```

[KISS](https://en.wikipedia.org/wiki/KISS_principle)!

In every bathroom in the world blue means cold,  red means hot.
When kids wants to draw a fire they use manly red not white.

My proposal, simple and intuitive for all humans that use hot/cold water at home 

Standby: Green (good to go)
Heat on < 60°C : BLUE! safe to touch!
Heat on > 60°C : RED! (do not touch it!)

only then we can discuss the below ones:

Overheat: RED flashing + buzzer! (do not touch it, ALARM!)
DFU upgrade: white
Error: yellow flashing! (means attention!)

smooth transition is for teenagers IMHO.

**Comment by g3gg0** [comment]
talking about a custom open sourced firmware for a mini microcontroller / OLED controlled USB-C PD hotplate and asking for KISS.
not the KISS i am used to, but thanks for sharing your thoughts ;)

**Comment by Ralim** [comment]
Hello,
Always open to new ideas.

Please note that DFU is NOT changeable, and thus is will always be blue and we won't use blue because of that.


The convention I went with is red/yellow== burn risk and green == safe. As red/green is far more used on tools than blue (blue is often an on light).

What would you mean by overheat?

**Comment by Zuckme** [comment]
> asking for KISS. not exactly the KISS i am used to, but thanks for sharing your thoughts ;)

Well if it is possible always go KISS!
I use [FAR Manager](https://www.farmanager.com/) in Windows... so believe me I love not KISS if it makes sense.

> Please note that DFU is NOT changeable, and thus is will always be blue and we won't use blue because of that.

To put the unit in DFU mode the user has to perform a not intuitive process.
It is impossible to kick the DFU in without knowing it and being 100% aware of that.
On top of that two things:
1) the display shows "DFU...."
2) DFU mode should not be used everyday. Only sporadic.

Long story short: it is impossible to confuse the blue DFU with the unit on and <60
 
> What would you mean by overheat?

Oh, I assumed there is a safety/emergency state when the unit is thermal drifitng with full power...
If there isn't ignore that. I got my MHP30 since two days....
That said all the color discussion boils down to my personal taste.

**Ralim** and **g3gg0** you did and do a wonderful job, and deserve all my respect.
I am sorry if I sound harsh and bossy.

As you already mentioned, colors must be user configurable somehow.
We can save a lot of discussion about personal tastes in this way.
